src/redprl/operator.sml

```diff
@@ -88,10 +88,86 @@ structure RedPrlParameterTerm = AbtParameterTerm (RedPrlParameter)
 
 structure RedPrlArity = ListAbtArity (structure PS = RedPrlParamSort and S = RedPrlSort)
 
+
+structure RedPrlKind =
+struct
+  (*
+   * DISCRETE < KAN < HCOM < PRE
+   *                < COE  <
+   *
+   * and KAN = meet (HCOM, COE)
+   *)
+  datatype kind = DISCRETE | KAN | HCOM | COE | PRE
+
+  val toString =
+    fn DISCRETE => "discrete"
+     | KAN => "kan"
+     | HCOM => "hcom"
+     | COE => "coe"
+     | PRE => "pre"
+
+  local
+    structure Internal :
+    (* this could be the new meet semi-lattice *)
+    sig
+      type t = kind
+      val top : t
+      val leq : t * t -> bool
+      val meet : t * t -> t
```

Comment:
I realized the current meet semi-lattice could be improved but I do not want to do the propagation again till I am sure what we really want.